### PR TITLE
Runner implementation

### PR DESCRIPTION
**Pull request by Eabramov84**

### CODE REVIEW

File: runner/runner.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+// StartInstance start service instance as systemd unit.
+func (runner *Runner) StartInstance(instanceID, runtimeDir string, params StartInstanceParams) (status InstanceStatus) {
+	status.InstanceID = instanceID
+	status.State = failedState
+
+	unitName := fmt.Sprintf(systemdUnitNameTamplate, instanceID)
+
+	runner.startingUnits[unitName] = make(chan dbus.UnitStatus)
```
Comment (al1img) [inline runner/runner.go:132]: Modification and read access to startingUnits map should be guarded with RW mutex.

File: runner/runner.go [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+		params.StartInterval = defaultStartTimout * time.Second
+	}
+
+	timeoutChannel := time.After(unitStartTimeoutMultiplier * params.StartInterval)
+
+	var currentState string
+
+	for {
+		select {
+		case unitStatus := <-runner.startingUnits[unitName]:
```
Comment (al1img) [inline runner/runner.go:249]: To avoid dead lock channel should be received as a parameter of this function.

File: runner/runner.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+// StartInstance start service instance as systemd unit.
+func (runner *Runner) StartInstance(instanceID, runtimeDir string, params StartInstanceParams) (status InstanceStatus) {
+	status.InstanceID = instanceID
+	status.State = failedState
+
+	unitName := fmt.Sprintf(systemdUnitNameTamplate, instanceID)
+
+	runner.startingUnits[unitName] = make(chan dbus.UnitStatus)
+
+	runner.subSet.Add(unitName)
```
Comment (al1img) [inline runner/runner.go:134]: It is not thread-safe to add/remove items to subSet after subscription. I'm afraid you should write your own subscription implementation based on SubscribeUnitsCustom.

File: runner/runner_test.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	"testing"
+	"time"
+
+	"github.com/aoscloud/aos_common/aoserrors"
+	"github.com/aoscloud/aos_servicemanager/runner"
+	"github.com/coreos/go-systemd/v22/dbus"
+	log "github.com/sirupsen/logrus"
+)
+
+func init() {
```
Comment (al1img) [inline runner/runner_test.go:34]: Add header before init:
```go
/******
 * Init
 ******
```

File: runner/runner_test.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+[Install]
+WantedBy=multi-user.target
+`
+
+const serviceContent = `
+#!/bin/bash
+echo "HELLO!!!!"
+sleep 10
+`
+const binFileName = "service.sh"
```
Comment (al1img) [inline runner/runner_test.go:71]: Why binFileName?

File: runner/runner_test.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+const runtimeDir = "/run/aos/runtime"
+
+const (
+	failedState = "failed"
```
Comment (al1img) [inline runner/runner_test.go:76]: Should be used from cloudprotocol consts.

File: runner/runner_test.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+	activeState = "active"
+)
+
+/***********************************************************************************************************************
```
Comment (al1img) [inline runner/runner_test.go:80]: Again init?

File: runner/runner_test.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+* Tests
+***********************************************************************************************************************/
+
+func TestStartStopService(t *testing.T) {
+	runnerInstance, err := runner.New()
+	if err != nil {
+		t.Fatalf("Can't create runner: %s", err)
+	}
+
```
Comment (al1img) [inline runner/runner_test.go:119]: defer runner.Close?

File: runner/runner_test.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+	if status.State != activeState {
+		t.Error("Service is not active")
+	}
+
+	// test no service binary
+	status = runnerInstance.StartInstance("someID", runtimeDir, runner.StartInstanceParams{StartInterval: 2 * time.Second})
+	if status.Err == nil {
+		t.Error("Should be error : ", runner.ErrStartInstance)
```
Comment (al1img) [inline runner/runner_test.go:142]: remove space befor :

File: runner/runner_test.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	if status.State != failedState {
+		t.Error("State should be failed ")
```
Comment (al1img) [inline runner/runner_test.go:146]: Remove space at the end.

File: runner/runner_test.go [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	serviceStatusesChannel := runnerInstance.InstanceStatusChannel()
+
+	// wait fro service failed
```
Comment (al1img) [inline runner/runner_test.go:151]: wait for

File: runner/runner_test.go [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+	failStatus := <-serviceStatusesChannel
+
+	if len(failStatus) != 1 {
+		t.Error("count of updated statuses should be 1")
```
Comment (al1img) [inline runner/runner_test.go:155]: Capital letter.

File: runner/runner_test.go [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+
+	if failStatus[0].InstanceID != "id1" {
+		t.Error("Incorrect instance id in status")
+	}
+
+	if failStatus[0].State != failedState {
+		t.Errorf("Incorrect service state: %s", failStatus[0].State)
+	}
+
+	// wait fro service failed
```
Comment (al1img) [inline runner/runner_test.go:166]: wait for

File: runner/runner_test.go [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+		t.Error("Incorrect instance id in status")
+	}
+
+	if activeStatus[0].State != activeState {
+		t.Errorf("Incorrect service state: %s", activeStatus[0].State)
+	}
+
+	// stop instnace
+	if err := runnerInstance.StopInstance("id1"); err != nil {
+		t.Error("Can't stop service: ", err)
```
Comment (al1img) [inline runner/runner_test.go:179]: Errorf %s

File: runner/runner_test.go [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	// test service not loaded
+	if err := runnerInstance.StopInstance("someID"); err != nil {
+		t.Error("Should no error")
```
Comment (al1img) [inline runner/runner_test.go:184]: Can't stop instance: %s, err

File: runner/runner_test.go [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+func setup() (err error) {
+	if err = os.MkdirAll(runtimeDir, 0o755); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	serviFile := path.Join(runtimeDir, "aos-service@.service")
```
Comment (al1img) [inline runner/runner_test.go:199]: serviceFile

File: runner/runner_test.go [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+		return aoserrors.Wrap(err)
+	}
+
+	defer systemd.Close()
+
+	if _, err = systemd.LinkUnitFilesContext(context.Background(), []string{serviFile}, true, true); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	err = systemd.ReloadContext(context.Background())
```
Comment (al1img) [inline runner/runner_test.go:216]: Combine in one line.

File: runner/aos-service@.service [hunk 18]
```diff
@@ -0,0 +1,20 @@
+[Unit]
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/aos-service@.service:1]: done

File: runner/aos-service@.service [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+After=network.target
+StartLimitIntervalSec=30
+StartLimitBurst=3
+
+[Service]
+Type=forking
+Restart=always
+RestartSec=1
+ExecStartPre=/usr/bin/{RUNNER} delete -f %i
+ExecStart=/usrt/bin/{RUNNER} run -d --pid-file /run/aos/runtime/%i/.pid -b /run/aos/runtime/%i %i
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/aos-service@.service:12]: renamed

File: runner/aos-service@.service [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+
+ExecStop=/usr/bin/{RUNNER} kill %i SIGKILL
+ExecStopPost=/usr/bin/{RUNNER} delete -f %i
+PIDFile=/run/aos/runtime/%i/.pid
+SuccessExitStatus=SIGKILL
+
+[Install]
+WantedBy=multi-user.target
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/aos-service@.service:20]: added

File: runner/runner.go [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+)
+
+/***********************************************************************************************************************
+ * Consts
+ **********************************************************************************************************************/
+
+const unitStatusChannelSize = 10
+
+const (
+	defaultStartTimout         = 30
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:39]: done

File: runner/runner.go [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+	unitStartTimeoutMultiplier = 2
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:40]: renamed

File: runner/runner.go [hunk 23]
```diff
@@ -0,0 +1,282 @@
...
+)
+
+const systemdUnitNameTamplate = "aos-service@%s.service"
+
+const errNotLoaded = "not loaded"
+
+const (
+	failedState   = "failed"
+	activeState   = "active"
+	jobeStateDone = "done"
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:50]: done

File: runner/runner.go [hunk 24]
```diff
@@ -0,0 +1,282 @@
...
+	InstanceID string
+	State      string
+	Err        error
+	ExitCode   int
+}
+
+// Runner runner instance.
+type Runner struct {
+	systemd            *dbus.Conn
+	subSet             *dbus.SubscriptionSet
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:81]: renamed to subscriptionSet

File: runner/runner.go [hunk 25]
```diff
@@ -0,0 +1,282 @@
...
+	channel := make(chan string)
+
+	if _, err := runner.systemd.StopUnitContext(context.Background(), unitName, "replace", channel); err != nil {
+		if strings.Contains(err.Error(), errNotLoaded) {
+			log.WithField("id", instanceID).Warn("Service not loaded")
+
+			return nil
+		}
+
+		log.WithField("id", instanceID).Errorf("Can't stop systemd unit: %s", err)
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:181]: changed to debug

File: runner/runner.go [hunk 26]
```diff
@@ -0,0 +1,282 @@
...
+
+	return nil
+}
+
+/***********************************************************************************************************************
+  Private
+ **********************************************************************************************************************/
+
+func (runner *Runner) monitorUnitSates() {
+	evChan, errChan := runner.subSet.Subscribe()
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:198]: renamed

File: runner/runner.go [hunk 27]
```diff
@@ -0,0 +1,282 @@
...
+
+	for {
+		select {
+		case changes := <-evChan:
+			unitsStatus := []InstanceStatus{}
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:203]: done

File: runner/runner.go [hunk 28]
```diff
@@ -0,0 +1,282 @@
...
+
+			for _, val := range changes {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:205]: done

File: runner/runner.go [hunk 29]
```diff
@@ -0,0 +1,282 @@
...
+				if val == nil {
+					continue
+				}
+
+				ch, ok := runner.startingUnits[val.Name]
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:210]: done

File: runner/runner.go [hunk 30]
```diff
@@ -0,0 +1,282 @@
...
+
+			if len(unitsStatus) == 0 {
+				continue
+			}
+
+			select {
+			case runner.instanceStatusChan <- unitsStatus:
+
+			default:
+				log.Info("Instance status channel full")
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:226]: prefer to have info.  It is not a critical scenarion

File: runner/runner.go [hunk 31]
```diff
@@ -0,0 +1,282 @@
...
+		}
+	}
+}
+
+func (runner *Runner) getStartingState(unitName string, params StartInstanceParams) string {
+	if params.StartInterval == 0 {
+		params.StartInterval = defaultStartTimout * time.Second
+	}
+
+	timeoutChannel := time.After(unitStartTimeoutMultiplier * params.StartInterval)
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:243]: done

File: runner/runner.go [hunk 32]
```diff
@@ -0,0 +1,282 @@
...
+
+			return activeState
+		}
+	}
+}
+
+func unitStatusToInstanceStatus(unitStatus *dbus.UnitStatus) (runnerStatus InstanceStatus) {
+	res := strings.ReplaceAll(unitStatus.Name, ".service", "")
+
+	runnerStatus.InstanceID = strings.ReplaceAll(res, "aos-service@", "")
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:269]: thanks

File: runner/runner.go [hunk 33]
```diff
@@ -0,0 +1,282 @@
...
+
+	runnerStatus.State = convertActiveStateToInstanceState(unitStatus.ActiveState)
+
+	return runnerStatus
+}
+
+func convertActiveStateToInstanceState(uintStatus string) string {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:276]: renamed

File: runner/runner_test.go (hunk 5 above)
Comment (Eabramov84) [reply runner/runner_test.go:61]: renamed to serviceFileName

File: runner/runner_test.go (hunk 7 above)
Comment (Eabramov84) [reply runner/runner_test.go:70]: removed

File: runner/runner.go (hunk 1 above)
Comment (Eabramov84) [reply runner/runner.go:136]: done

File: runner/runner.go (hunk 3 above)
Comment (Eabramov84) [reply runner/runner.go]: reworked

File: runner/runner_test.go (hunk 6 above)
Comment (Eabramov84) [reply runner/runner_test.go]: done

File: runner/runner.go [hunk 34]
```diff
@@ -0,0 +1,282 @@
...
+	defaultStartTimout         = 30
+	unitStartTimeoutMultiplier = 2
+)
+
+const systemdUnitNameTamplate = "aos-service@%s.service"
+
+const errNotLoaded = "not loaded"
+
+const (
+	failedState   = "failed"
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go]: done

File: runner/runner.go [hunk 35]
```diff
@@ -0,0 +1,282 @@
...
+		startingUnits:      make(map[string]chan dbus.UnitStatus),
+		stopChan:           make(chan bool, 1),
+	}
+
+	// Create systemd connection
+	runner.systemd, err = dbus.NewSystemConnectionContext(context.Background())
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:104]: this approach doesn't work for d-bus api

File: runner/aos-service@.service [hunk 36]
```diff
@@ -0,0 +1,20 @@
...
+After=network.target
+StartLimitIntervalSec=30
+StartLimitBurst=3
+
+[Service]
+Type=forking
+Restart=always
+RestartSec=1
+ExecStartPre=/usr/bin/{RUNNER} delete -f %i
+ExecStart=/usrt/bin/{RUNNER} run -d --pid-file /var/aos/runtime/%i/.pid -b /var/aos/runtime/%i %i
```
Comment (al1img) [inline runner/aos-service@.service:12]: In all  cases /var/aos/runtime -> /run/aos/runtime

File: runner/runner.go [hunk 37]
```diff
@@ -0,0 +1,316 @@
...
+	stopChan           chan bool
+	sync.RWMutex
+}
+
+/***********************************************************************************************************************
+  Public
+ **********************************************************************************************************************/
+
+// New creates new systemd runner.
+func New() (runner *Runner, err error) {
```
Comment (al1img) [inline runner/runner.go:91]: You can remove named values from return.

File: runner/runner.go [hunk 38]
```diff
@@ -0,0 +1,316 @@
...
+
+	runner.systemd.Close()
+}
+
+func (runner *Runner) InstanceStatusChannel() <-chan []InstanceStatus {
+	return runner.instanceStatusChan
+}
+
+// StartInstance starts service instance as systemd unit.
+func (runner *Runner) StartInstance(instanceID, runtimeDir string, params StartInstanceParams) (status InstanceStatus) {
```
Comment (al1img) [inline runner/runner.go:124]: Remove named value from return.

File: runner/runner.go [hunk 39]
```diff
@@ -0,0 +1,316 @@
...
+	Err        error
+	ExitCode   int
+}
+
+// Runner runner instance.
+type Runner struct {
+	systemd            *dbus.Conn
+	instanceStatusChan chan []InstanceStatus
+	startingUnits      map[string]chan dbus.UnitStatus
+	runningUnits       map[string]bool
```
Comment (al1img) [inline runner/runner.go:81]: you can handle the same functionality with one map: when service is started assign channel value to nil.

File: runner/runner.go [hunk 40]
```diff
@@ -0,0 +1,316 @@
...
+	log.WithFields(log.Fields{"id": instanceID, "status": status}).Debug("Stop service")
+
+	return nil
+}
+
+/***********************************************************************************************************************
+  Private
+ **********************************************************************************************************************/
+
+func (runner *Runner) monitorUnitSates() {
```
Comment (al1img) [inline runner/runner.go:209]: States

File: runner/runner.go [hunk 41]
```diff
@@ -0,0 +1,282 @@
...
+	failedState   = "failed"
+	activeState   = "active"
+	jobeStateDone = "done"
+)
+
+/***********************************************************************************************************************
+ * Vars
+ **********************************************************************************************************************/
+
+var ErrStartInstance = errors.New("can't start service instance")
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:55]: removed

File: runner/runner.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:243]: changed to error

File: runner/aos-service@.service (hunk 36 above)
Comment (Eabramov84) [reply runner/aos-service@.service:12]: fixed

File: runner/runner.go [hunk 42]
```diff
@@ -0,0 +1,316 @@
...
+}
+
+// Runner runner instance.
+type Runner struct {
+	systemd            *dbus.Conn
+	instanceStatusChan chan []InstanceStatus
+	startingUnits      map[string]chan dbus.UnitStatus
+	runningUnits       map[string]bool
+	stopChan           chan bool
+	sync.RWMutex
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:83]: done

File: runner/runner.go (hunk 37 above)
Comment (Eabramov84) [reply runner/runner.go:91]: in this particular case prefer to have named return values

File: runner/runner.go [hunk 43]
```diff
@@ -0,0 +1,316 @@
...
+	ExitCode   int
+}
+
+// Runner runner instance.
+type Runner struct {
+	systemd            *dbus.Conn
+	instanceStatusChan chan []InstanceStatus
+	startingUnits      map[string]chan dbus.UnitStatus
+	runningUnits       map[string]bool
+	stopChan           chan bool
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:82]: done

File: runner/runner.go (hunk 38 above)
Comment (Eabramov84) [reply runner/runner.go:124]: removed

File: runner/runner.go (hunk 39 above)
Comment (Eabramov84) [reply runner/runner.go:81]: reworked

File: runner/runner.go (hunk 40 above)
Comment (Eabramov84) [reply runner/runner.go:209]: done

File: runner/runner.go [hunk 44]
```diff
@@ -0,0 +1,316 @@
...
+	}
+
+	return cloudprotocol.InstanceStateFailed
+}
+
+// mismatchUnitStatus returns true if the provided UnitStatus objects
+// are not equivalent. false is returned if the objects are equivalent.
+// Only the Name, Description and state-related fields are used in
+// the comparison.
+func mismatchUnitStatus(u1, u2 *dbus.UnitStatus) bool {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:310]: done

File: runner/runner_test.go [hunk 45]
```diff
@@ -0,0 +1,211 @@
...
+	"context"
+	"io/ioutil"
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/aoscloud/aos_common/aoserrors"
+	"github.com/aoscloud/aos_common/api/cloudprotocol"
+	"github.com/aoscloud/aos_servicemanager/runner"
```
Comment (al1img) [inline runner/runner_test.go:30]: Move to separate local block.

File: runner/runner_test.go [hunk 46]
```diff
@@ -0,0 +1,211 @@
...
+`
+
+const serviceContent = `
+#!/bin/bash
+echo "HELLO!!!!"
+sleep 10
+`
+const serviceFileName = "service.sh"
+
+const runtimeDir = "/var/aos/runtime"
```
Comment (al1img) [inline runner/runner_test.go:64]: `/run/aos/runtime`

File: runner/runner_test.go [hunk 47]
```diff
@@ -0,0 +1,211 @@
...
+/***********************************************************************************************************************
+* Tests
+***********************************************************************************************************************/
+
+func TestStartStopService(t *testing.T) {
+	runnerInstance, err := runner.New()
+	if err != nil {
+		t.Fatalf("Can't create runner: %s", err)
+	}
+
```
Comment (al1img) [inline runner/runner_test.go:105]: remove empty line.

File: runner/runner.go [hunk 48]
```diff
@@ -0,0 +1,316 @@
...
+	channel := make(chan string)
+
+	if _, err := runner.systemd.StopUnitContext(context.Background(), unitName, "replace", channel); err != nil {
+		if strings.Contains(err.Error(), errNotLoaded) {
+			log.WithField("id", instanceID).Warn("Service not loaded")
+
+			return nil
+		}
+
+		log.WithField("id", instanceID).Debugf("Can't stop systemd unit: %s", err)
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:185]: done

File: runner/aos-service@.service [hunk 49]
```diff
@@ -0,0 +1,20 @@
...
+Description=AOS Service
+After=network.target
+StartLimitIntervalSec=30
+StartLimitBurst=3
+
+[Service]
+Type=forking
+Restart=always
+RestartSec=1
+ExecStartPre=/usr/bin/{RUNNER} delete -f %i
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/aos-service@.service:11]: it replaces while yocto build

File: runner/runner.go [hunk 50]
```diff
@@ -0,0 +1,310 @@
...
+const unitStatusChannelSize = 10
+
+const (
+	defaultStartTimeout    = 30
+	startTimeoutMultiplier = 2
+)
+
+const systemdUnitNameTamplate = "aos-service@%s.service"
+
+const errNotLoaded = "not loaded"
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:46]: done

File: runner/runner.go [hunk 51]
```diff
@@ -0,0 +1,310 @@
...
+// New creates new systemd runner.
+func New() (runner *Runner, err error) {
+	runner = &Runner{
+		instanceStatusChan: make(chan []InstanceStatus, unitStatusChannelSize),
+		runningUnits:       make(map[string]chan dbus.UnitStatus),
+		stopChan:           make(chan struct{}, 1),
+	}
+
+	// Create systemd connection
+	runner.systemd, err = dbus.NewSystemConnectionContext(context.Background())
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply runner/runner.go:91]: done

File: runner/runner_test.go (hunk 45 above)
Comment (Eabramov84) [reply runner/runner_test.go:30]: done

File: runner/runner_test.go (hunk 46 above)
Comment (Eabramov84) [reply runner/runner_test.go:64]: done

File: runner/runner_test.go (hunk 47 above)
Comment (Eabramov84) [reply runner/runner_test.go:105]: done

**Review by al1img — APPROVED** [review]

**Pull request merged**